cEP-0026.md

```diff
@@ -0,0 +1,189 @@
+
+| Metadata |                                           |
+| -------- | ----------------------------------------- |
+| cEP      | 26                                        |
+| Version  | 1.0                                       |
+| Title    | Optimize caching for the NextGen-Core     |
+| Authors  | Palash Nigam <mailto:npalash25@gmail.com> |
+| Status   | Implementation Due                        |
+| Type     | Feature                                   |
+
+## Abstract
+
+This document describes how a caching mechanism (an enhanced version of the one
+being used by the old core) will be implemented and integrated with the
+NextGen-Core as a part of this
+[GSoC project](https://summerofcode.withgoogle.com/projects/#6434190552203264).
+
+## Introduction
+
+Since the NextGen-Core aims to provide faster speed and better performance for
+coala-runs on large codebases it needs caching to achieve this kind of
+performance.
+
+Structure wise the NextGen cache is a dictionary with bear types and cache
+tables as key-value pairs. The cache tables themselves are
+dictionary-like-objects that map the `persistent_hash`s of the task objects to
+the bear results. This current mechanism will be further enhanced by adding
+features like `FileFactory` class to load files, `Directory` class to interface
+with the directories, ignoring unmodified directories during successive runs and
+adding cache control flags for better usability and providing the user various
+options to perform caching according to a project's needs.
+
+## Implementation
+
+The implementation will be divided into four parts
+
+- Prototype for `FileFactory` class implementation.
+- Prototype for `Directory` class implementation.
+- Ignore directories functionality.
+- Cache control flags.
+
+### Prototype of FileFactory class implementation
+
+This class will follow the Factory design pattern and will act as a proxy to
+dealing with files. This class will have various methods to provide the file
+contents in different forms like string, binary format, etc. With the help of
+this class that we can replace the actual file contents in our file-dict with
+these objects which will itself be a performance improvement.
+
+The following is the prototype for `FileFactory` class.
+
+```python
+import linecache
+import os
+
+from coalib.core.PersistentHash import persistent_hash
+
+
+class FileFactory:
+    def __init__(self, filename, parent):
+        self.filename = filename
+        self.parent = os.path.abspath(os.path.join(self.filename, os.pardir))
+
+    @classmethod
+    def open(cls, filename, newline=None):
+        return cls(filename, newline=newline)
+
+    def file_as_bytes(self):
+        with open(self.filename, 'rb') as fp:
+            return fp.read()
+
+    def file_as_string(self):
+        return self.file_as_bytes().decode()
+
+    def file_as_list(self):
+        self.lines = linecache.getlines(self.filename)
+        self.linecount = len(self.lines)
```

Comment:
hm a bit inconsistent that this function suddenly uses `linecache` while the others don't ;) but not so crucial right now.